File: src/loopflow/lf/ops/trace.py
"""Ops tracing for parity testing.

When LF_TRACE=1, ops commands emit JSON traces instead of executing.
This enables comparison between Python and Rust implementations.
"""

from __future__ import annotations

import hashlib
import json
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


def trace_enabled() -> bool:
    """Check if tracing is enabled via LF_TRACE env var."""
    return os.environ.get("LF_TRACE", "").lower() in ("1", "true", "yes")


@dataclass
class OpTrace:
    """A single traced operation."""

    op: str
    result: str | None = None
    args: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {"op": self.op}
        if self.result is not None:
            d["result"] = self.result
        if self.args:
            d.update(self.args)
        return d


@dataclass
class Tracer:
    """Collects operation traces during ops command execution."""

    command: str
    options: dict[str, Any] = field(default_factory=dict)
    operations: list[OpTrace] = field(default_factory=list)

    def trace(self, op: str, result: str | None = None, **kwargs: Any) -> None:
        """Record an operation."""
        self.operations.append(OpTrace(op=op, result=result, args=kwargs))

    def to_dict(self) -> dict[str, Any]:
        return {
            "command": self.command,
            "options": self.options,
            "operations": [op.to_dict() for op in self.operations],
        }

    def to_json(self) -> str:
        return json.dumps(self.to_dict(), indent=2)


def hash_prompt(prompt: str) -> str:
    """Hash a prompt for trace comparison (avoids large diffs)."""
    return hashlib.sha256(prompt.encode()).hexdigest()[:16]


def normalize_path(path: Path | str, repo_root: Path) -> str:
    """Normalize a path to be relative to repo root."""
    try:
        return str(Path(path).relative_to(repo_root))
    except ValueError:
        return str(path)


# Mock responses for trace mode
MOCK_RESPONSES = {
    "git:status": "dirty",
    "git:diff_cached": "has_changes",
    "git:has_upstream": True,
    "lint:run": "pass",
    "agent:commit_message": {"title": "mock commit", "body": ""},
    "agent:pr_message": {"title": "mock pr", "body": ""},
    "gh:pr_exists": False,
}


def mock_response(op: str) -> Any:
    """Get a mock response for an operation in trace mode."""
    return MOCK_RESPONSES.get(op)

File: rust/loopflow/src/ops/trace.rs
//! Ops tracing for parity testing.
//!
//! When LF_TRACE=1, ops commands emit JSON traces instead of executing.

use serde::Serialize;
use sha2::{Digest, Sha256};
use std::env;

/// Check if tracing is enabled via LF_TRACE env var.
pub fn trace_enabled() -> bool {
    env::var("LF_TRACE")
        .map(|v| matches!(v.to_lowercase().as_str(), "1" | "true" | "yes"))
        .unwrap_or(false)
}

/// A single traced operation.
#[derive(Debug, Clone, Serialize)]
pub struct OpTrace {
    pub op: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<String>,
    #[serde(flatten)]
    pub args: serde_json::Value,
}

impl OpTrace {
    pub fn new(op: &str) -> Self {
        Self {
            op: op.to_string(),
            result: None,
            args: serde_json::Value::Null,
        }
    }

    pub fn with_result(mut self, result: &str) -> Self {
        self.result = Some(result.to_string());
        self
    }

    pub fn with_args(mut self, args: serde_json::Value) -> Self {
        self.args = args;
        self
    }
}

/// Collects operation traces during ops command execution.
#[derive(Debug, Clone, Serialize)]
pub struct Tracer {
    pub command: String,
    pub options: serde_json::Value,
    pub operations: Vec<OpTrace>,
}

impl Tracer {
    pub fn new(command: &str, options: serde_json::Value) -> Self {
        Self {
            command: command.to_string(),
            options,
            operations: Vec::new(),
        }
    }

    pub fn trace(&mut self, op: &str) {
        self.operations.push(OpTrace::new(op));
    }

    pub fn trace_result(&mut self, op: &str, result: &str) {
        self.operations.push(OpTrace::new(op).with_result(result));
    }

    pub fn trace_args(&mut self, op: &str, args: serde_json::Value) {
        self.operations.push(OpTrace::new(op).with_args(args));
    }

    pub fn trace_full(&mut self, op: &str, result: Option<&str>, args: serde_json::Value) {
        let mut trace = OpTrace::new(op).with_args(args);
        if let Some(r) = result {
            trace = trace.with_result(r);
        }
        self.operations.push(trace);
    }

    pub fn to_json(&self) -> String {
        serde_json::to_string_pretty(self).expect("tracer serialization")
    }
}

/// Hash a prompt for trace comparison.
pub fn hash_prompt(prompt: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(prompt.as_bytes());
    let result = hasher.finalize();
    hex::encode(&result[..8])
}

/// Mock responses for trace mode.
pub struct MockResponses;

impl MockResponses {
    pub fn git_status() -> &'static str {
        "dirty"
    }

    pub fn git_diff_cached() -> &'static str {
        "has_changes"
    }

    pub fn git_has_upstream() -> bool {
        true
    }

    pub fn gh_pr_exists() -> bool {
        false
    }
}
